core/trino-main/src/main/java/io/trino/connector/WorkerDynamicCatalogManager.java

```diff
@@ -128,22 +144,28 @@ public void pruneCatalogs(Set<CatalogHandle> catalogsInUse)
             }
         }
         finally {
-            catalogsUpdateLock.unlock();
+            catalogCleanUpLock.unlock();
         }
 
-        // todo do this in a background thread
-        for (CatalogConnector removedCatalog : removedCatalogs) {
-            try {
-                removedCatalog.shutdown();
-            }
-            catch (Throwable e) {
-                log.error(e, "Error shutting down catalog: %s".formatted(removedCatalog));
-            }
-        }
-        if (!removedCatalogs.isEmpty()) {
-            List<String> sortedHandles = removedCatalogs.stream().map(connector -> connector.getCatalogHandle().toString()).sorted().toList();
-            log.debug("Pruned catalogs: %s", sortedHandles);
-        }
+        var ignored = Futures.whenAllComplete(removedCatalogs.stream()
+                .map(removedCatalog -> Futures.submit(
+                        () -> {
+                            try {
+                                removedCatalog.shutdown();
+                            }
+                            catch (Throwable e) {
+                                log.error(e, "Error shutting down catalog: %s".formatted(removedCatalog));
+                            }
+                        },
+                        catalogWorkService))
+                .collect(toImmutableList()))
+                .run(() ->
```

Comment:
This is hard to read with the multiple levels of nesting. For instance, it's not immediately clear that that `.run()` is invoked on the result of `whenAllComplete()`. Visually, it looks as if it's chained after `.collect()`.

Reply:
I'll make a future out of `whenAllComplete` with a relevant name and then execute the run on that?